File: 134.GasStation.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        n, start, end, S = len(gas), 0, 0, 0
        while True:
            if S + gas[end] < cost[end]:
                if end == start:
                    start += 1
                    end = start
                    S = 0
                else:
                    S += cost[start] - gas[start]
                    start += 1
                if start == n: 
                    return -1
            else:
                S += gas[end] - cost[end]
                end += 1
                if end == n:
                    end = 0
                if end == start:
                    return start

File: Algorithms/cpp/134.GasStation.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        int start = 0;
        int sum = 0;
        int end = start;
        while (true) {
            if (sum + gas[end] < cost[end]) {
                if (end == start) {
                    start ++;
                    end = start;
                    sum = 0;
                } else {
                    sum += cost[start] - gas[start];
                    start ++;
                }
                if (start == n) {
                    return -1;
                }
            } else {
                sum += gas[end] - cost[end];
                end ++;
                if (end == n) {
                    end = 0;
                }
                if (end == start) {
                    return start;
                }
            }
        }
    }
};
